src/main/java/org/graylog/aws/inputs/transports/KinesisTransport.java

```diff
@@ -44,38 +54,98 @@
+
+    public static final int DEFAULT_THROTTLED_WAIT = 60000;
+    public static final int DEFAULT_BATCH_SIZE = 10000;
+    public static final int KINESIS_CONSUMER_STOP_WAIT_MS = 15000;
 
     private final Configuration configuration;
     private final org.graylog2.Configuration graylogConfiguration;
     private final NodeId nodeId;
     private final LocalMetricRegistry localRegistry;
     private final ClusterConfigService clusterConfigService;
+    private final ObjectMapper objectMapper;
 
     private KinesisConsumer reader;
+    ExecutorService executor = null;
+    public Future<?> kinesisTaskFuture = null;
+
+    /**
+     * Indicates if the Kinesis consumer has been stopped due to throttling. Allows the consumer to be restarted
+     * once throttling is cleared.
+     */
+    public AtomicBoolean stoppedDueToThrottling = new AtomicBoolean(false);
+    public KinesisTransportState consumerState = STOPPED;
 
     @Inject
     public KinesisTransport(@Assisted final Configuration configuration,
+                            EventBus serverEventBus,
                             org.graylog2.Configuration graylogConfiguration,
                             final ClusterConfigService clusterConfigService,
                             final NodeId nodeId,
-                            LocalMetricRegistry localRegistry) {
+                            LocalMetricRegistry localRegistry,
+                            @AWSObjectMapper ObjectMapper objectMapper) {
+        super(serverEventBus, configuration);
         this.clusterConfigService = clusterConfigService;
         this.configuration = configuration;
         this.graylogConfiguration = graylogConfiguration;
         this.nodeId = nodeId;
         this.localRegistry = localRegistry;
+        this.objectMapper = objectMapper;
     }
 
     @Override
-    public void launch(MessageInput input) throws MisfireException {
-        ExecutorService executor = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
-                .setDaemon(true)
-                .setNameFormat("aws-kinesis-reader-%d")
-                .setUncaughtExceptionHandler((t, e) -> LOG.error("Uncaught exception in AWS Kinesis reader.", e))
-                .build());
+    public void handleChangedThrottledState(boolean isThrottled) {
+
+        if (!isThrottled) {
+            LOG.info("✅ Unthrottled");
+        } else {
+            LOG.info("❗Throttled");
+        }
+
+        if (!isThrottled && stoppedDueToThrottling.get()) {
+
+            stoppedDueToThrottling.set(false);
+
+            LOG.debug("Transport state [{}]", consumerState);
+
+            switch (consumerState) {
+                case STOPPED:
+                    LOG.info("[unthrottled] Throttle state ended restarting consumer");
+                    restartConsumer();
+                    break;
+                case STOPPING: {
+
+                    LOG.info("Kinesis consumer is still stopping. Waiting [{}ms] for the consumer to finish " +
```

Comment:
Improve this logic. Check if there is a built-in timeout in the stoppage of the AWS Kinesis consumer client. If so, repeatedly retry once/s with max wait time (or consider using exponential backoff).

Reply:
Ok as is.